python/cudf/cudf/core/udf/groupby_function.py

```diff
@@ -0,0 +1,867 @@
+# Copyright (c) 2020-2022, NVIDIA CORPORATION.
+
+import glob
+import math
+import os
+
+import cupy as cp
+import numba
+import numpy as np
+from numba import cuda, types
+from numba.core import cgutils
+from numba.core.extending import (
+    lower_builtin,
+    make_attribute_wrapper,
+    models,
+    register_model,
+    type_callable,
+    typeof_impl,
+)
+from numba.core.typing import signature as nb_signature
+from numba.core.typing.templates import AbstractTemplate, AttributeTemplate
+from numba.cuda.cudadecl import registry as cuda_registry
+from numba.cuda.cudadrv.devices import get_context
+from numba.cuda.cudaimpl import lower as cuda_lower
+from numba.np import numpy_support
+from numba.types import Record
+
+from cudf.core.column import as_column
+from cudf.core.udf.templates import (
+    group_initializer_template,
+    groupby_apply_kernel_template,
+)
+from cudf.core.udf.utils import (
+    _all_dtypes_from_frame,
+    _compile_or_get,
+    _get_kernel_groupby_apply,
+    _get_udf_return_type,
+    _supported_cols_from_frame,
+    _supported_dtypes_from_frame,
+)
+from cudf.utils.utils import _cudf_nvtx_annotate
+
+# Disable occupancy warnings to avoid polluting output when there are few
+# groups.
+numba.config.CUDA_LOW_OCCUPANCY_WARNINGS = 0
+
+
+class Group(object):
+    def __init__(self, group_data, size, index, dtype, index_dtype):
+        self.group_data = group_data
+        self.size = size
+        self.index = index
+        self.dtype = dtype
+        self.index_dtype = index_dtype
+
+
+class GroupType(numba.types.Type):
+    def __init__(self, group_scalar_type, index_type=types.int64):
+        self.group_scalar_type = group_scalar_type
+        self.index_type = index_type
+        self.group_data_type = types.CPointer(group_scalar_type)
+        self.size_type = types.int64
+        self.group_index_type = types.CPointer(index_type)
+        super().__init__(
+            name=f"Group({self.group_scalar_type}, {self.index_type})"
+        )
+
+
+@typeof_impl.register(Group)
+def typeof_group(val, c):
+    return GroupType(
+        numba.np.numpy_support.from_dtype(val.dtype),
+        numba.np.numpy_support.from_dtype(val.index_dtype),
+    )  # converting from numpy type to numba type
+
+
+@type_callable(Group)
+def type_group(context):
```

Comment:
I think it's worth adding a comment here explaining that this refers to the typing of the python "function" `Group.__init__` as it appears in python code, or renaming this function to `group_constructor`